### PR TITLE
Added MCU support for ArteryTek AT32F415

### PR DESCRIPTION
**Pull request by HorrorTroll**
## Description

This PR is added support for AT32F415 to QMK.

### Driver:
<details>
  <summary>Click to expand!</summary>

|      System      |                                      Support                                     |
| ---------------- | -------------------------------------------------------------------------------- |
| ADC driver       | :heavy_check_mark:                                                               |
| Audio            | :heavy_check_mark: (using PWM software or hardware, since MCU don't support DAC) |
| Backlight        | :heavy_check_mark: (using PWM, Timer and Software)                               |
| I2C driver       | :heavy_check_mark:                                                               |
| SPI driver       | :heavy_check_mark:                                                               |
| WS2812 driver    | :heavy_check_mark: (using Bitbang, SPI or PWM driver)                            |
| External EEPROMs | :heavy_multiplication_x:                                                         |
| EEPROM emulation | :heavy_check_mark:                                                               |
| serial driver    | :heavy_check_mark: (using Full-duplex (1))                                       |
| UART driver      | :exclamation: (already tested on ChibiOS code, should be work here too)          |

#### Notes:
- (1): Half-duplex is not tested.
- :heavy_multiplication_x: is meant for not testing either, since I don't have any hardware to try them.
</details>

### Handwired Onekey:
<details>
  <summary>Click to expand!</summary>

|        Keymap        |                   Testing                   |
| -------------------- | ------------------------------------------- |
| adc                  | :heavy_check_mark:                          |
| apa102               | :heavy_multiplication_x:                    |
| backlight            | :heavy_check_mark:                          |
| chibios_waiting_test | :exclamation: (need Logic Analyzer)         |
| console              | :heavy_check_mark:                          |
| default              | :heavy_check_mark:                          |
| digitizer            | :heavy_multiplication_x:                    |
| dip_switch_map       | :heavy_multiplication_x:                    |
| eep_rst              | :heavy_check_mark:                          |
| haptic               | :heavy_multiplication_x:                    |
| hardware_id          | :heavy_check_mark:                          |
| i2c_scanner          | :heavy_check_mark:                          |
| joystick             | :heavy_multiplication_x:                    |
| lvgl                 | :heavy_multiplication_x: (not enough flash) |
| oled                 | :heavy_check_mark:                          |
| quine                | :heavy_check_mark:                          |
| reboot               | :heavy_check_mark:                          |
| reset                | :heavy_check_mark:                          |
| rgb                  | :heavy_check_mark:                          |
| wear_leveling        | :heavy_check_mark:                          |

#### Notes:
- :heavy_multiplication_x: is meant for not testing, since I don't have any hardware to try them.
</details>

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by HorrorTroll** [comment]
Still on hold for waiting preview, thanks.

**Comment by tzarc** [comment]
Have asked a couple of members of the team to pick up AT-START-F415's for validation purposes.

File: docs/flashing.md
Comment (zvecr) [inline docs/flashing.md:361]: Toolbox looks for known pid:vid pairs when flashing, and no PR has been made to add support, so this statement here is incorrect.

File: platforms/chibios/drivers/serial_usart.c
Comment (zvecr) [inline platforms/chibios/drivers/serial_usart.c:17]: I cant see support for SIO for this chip, so this else doesnt seem correct?

Comment (HorrorTroll) [reply platforms/chibios/drivers/serial_usart.c]: Yeah, probably I just forgot to check and slip out of it for other chip needed. Already push a new commit to resolved this. Thanks!

File: docs/flashing.md
Comment (HorrorTroll) [reply docs/flashing.md]: I've just removed, will added back when Toolbox supported.

**Comment by MHooijberg** [comment]
Hi,

Is there anything I can do to help speed up the process, maybe work making the QMK CI Build action pass?
Or is this PR currently just waiting on review?

Thanks.

**Comment by HorrorTroll** [comment]
> Hi,
> 
> Is there anything I can do to help speed up the process, maybe work making the QMK CI Build action pass? Or is this PR currently just waiting on review?
> 
> Thanks.

Hi, I really appreciate that you want for help. I was reading your mail recently.

Right now the PR is waiting for review from QMK collaboration, so it might take time for them.

**Comment by MHooijberg** [comment]
> > Hi,
> > Is there anything I can do to help speed up the process, maybe work making the QMK CI Build action pass? Or is this PR currently just waiting on review?
> > Thanks.
> 
> Hi, I really appreciate that you want for help. I was reading your mail recently.
> 
> Right now the PR is waiting for review from QMK collaboration, so it might take time for them.

Alright,
if you do need anything, I have a keyboard that uses this chip.

I'll wait on QMK and start working on a keyboard branch then.
Thanks again, also thanks for all the work done for adding this MCU family.

**Comment by drashna** [comment]
A few of the one-key builds are failing.  While not critical, it would be nice to have them passing. 

* APA102 config is wrong/missing.
* SOLENOID_PIN(S) isn't defined
* SPI peripheral not configured for lvgl keymap

**Review by tzarc — APPROVED** [review]

**Pull request merged**
